keel-igor/src/main/kotlin/com/netflix/spinnaker/keel/igor/BuildLifecycleMonitor.kt

```diff
@@ -0,0 +1,197 @@
+package com.netflix.spinnaker.keel.igor
+
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.fasterxml.jackson.module.kotlin.convertValue
+import com.netflix.spinnaker.config.LifecycleConfig
+import com.netflix.spinnaker.keel.api.artifacts.BuildMetadata
+import com.netflix.spinnaker.keel.front50.Front50Service
+import com.netflix.spinnaker.keel.lifecycle.LifecycleEventStatus.FAILED
+import com.netflix.spinnaker.keel.lifecycle.LifecycleEventStatus.RUNNING
+import com.netflix.spinnaker.keel.lifecycle.LifecycleEventStatus.SUCCEEDED
+import com.netflix.spinnaker.keel.lifecycle.LifecycleEventStatus.UNKNOWN
+import com.netflix.spinnaker.keel.lifecycle.LifecycleEventType
+import com.netflix.spinnaker.keel.lifecycle.LifecycleEventType.BUILD
+import com.netflix.spinnaker.keel.lifecycle.LifecycleMonitor
+import com.netflix.spinnaker.keel.lifecycle.LifecycleMonitorRepository
+import com.netflix.spinnaker.keel.lifecycle.MonitoredTask
+import com.netflix.spinnaker.keel.igor.artifact.ArtifactMetadataService
+import com.netflix.spinnaker.keel.lifecycle.LifecycleEventStatus.ABORTED
+import com.netflix.spinnaker.keel.retrofit.isNotFound
+import kotlinx.coroutines.runBlocking
+import org.slf4j.LoggerFactory
+import org.springframework.beans.factory.annotation.Value
+import org.springframework.boot.context.properties.EnableConfigurationProperties
+import org.springframework.context.ApplicationEventPublisher
+import org.springframework.stereotype.Component
+import retrofit2.HttpException
+
+@Component
+@EnableConfigurationProperties(LifecycleConfig::class)
+class BuildLifecycleMonitor(
+  override val monitorRepository: LifecycleMonitorRepository,
+  override val publisher: ApplicationEventPublisher,
+  override val lifecycleConfig: LifecycleConfig,
+  val objectMapper: ObjectMapper,
+  val artifactMetadataService: ArtifactMetadataService,
+  val front50Service: Front50Service,
+  @Value("\${spinnaker.baseUrl}") private val spinnakerBaseUrl: String
+) : LifecycleMonitor(monitorRepository, publisher, lifecycleConfig) {
+  private val log by lazy { LoggerFactory.getLogger(javaClass) }
+
+  override fun handles(type: LifecycleEventType): Boolean =
+    type == BUILD
+
+  override suspend fun monitor(task: MonitoredTask) {
+    val buildData = parseAndValidate(task) ?: return
+    kotlin.runCatching {
```

Comment:
is calling to `getArtifactMetadata` done only if the build data was missing?

Reply:
No, it's how we update the status. so it's called every 30 seconds for any active builds. Likely it'll be called max 4 times for each build, since they're mostly done when we get them.

Should we call a different endpoint instead?

Reply:
I'm thinking if we could utilize the info we are currently storing in the DB somehow like only call this endpoint if the build is in `building` status.

Reply:
Hmm, good point. Let me do that!

Reply:
@gal-yardeni that's a good idea, I'll do that.

Reply:
Actually, I want to leave the logic that encapsulates status in one place. Right now I can't think of a good way to publish the right events w/o checking the API, or wiring in the other repository (which isn't super easy because we don't have the full delivery artifact, cause saving that was hard). I think I'm going to leave it as is, and if the load gets too high on the API we're calling we can update it. does that sound okay?

Reply:
Good point. I just looked at your code again, and another thing you can do is to pass the build status in `ArtifactListener`:`createBuildLifecycleEvent`. If it's already in a terminal status, you could save that API call.

Reply:
@gal-yardeni ya, but then that does weird things with how I roll up the events into user-facing steps (like, how do you get both the start and end time?). I've added a todo to make this a little better.

Reply:
you can pass the entire build metadata for it, I think. The reason I'm suggesting it is because if the build is terminal, the build metadata object you'll get from this call is duplicate to the one in `artifactListener`, and as we have some precentege of builds already in terminal states, we don't need to monitor them again, just send an event.
We should give rocket a heads up that we are adding another scheduled call to their services, just in case :)

Reply:
+1 to Gal's suggestion. Also, in my observation, we rarely ever go back to fix `TODOs` in the code...